src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/SessionContextValueResolver.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\MapSessionContext;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+
+final class SessionContextValueResolver implements ValueResolverInterface
+{
+    public function resolve(Request $request, ArgumentMetadata $argument): array
+    {
+        if (!$request->hasSession()) {
+            return [];
+        }
+
+        $attribute = $argument->getAttributesOfType(MapSessionContext::class, ArgumentMetadata::IS_INSTANCEOF)[0]
+            ?? null;
+
+        if (!$attribute) {
+            return [];
+        }
+
+        $type = $argument->getType();
+        if (null === $type) {
+            throw new \LogicException(sprintf('#[MapSessionContext] cannot be used on controller argument "$%s" with no type; a class must be provided.', $argument->getName()));
+        }
+        if (!class_exists($type)) {
+            throw new \LogicException(sprintf('#[MapSessionContext] cannot be used on controller argument "$%s" of type "%s"; a class must be provided.', $argument->getName(), $type));
+        }
+
+        if ($request->getSession()->has($type)) {
+            return [
+                $request->getSession()->get($type),
+            ];
+        }
+
+        $object = new $type();
```

Comment:
unless the argument accepts null, in which case we should provide null IMHO (allows for custom instantiation when the constructor has arguments)